in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+// once this tx has gone through.
+await this.wallet.registerContract({ artifact: this.artifact, instance: this.getInstance(options) });
+
+// ...
+
+if (options.estimateGas) {
+  // Why do we call this seemingly idempotent getter method here, without using its return value?
+  // This call pushes a capsule required for contract class registration under the hood. And since
+  // capsules are a stack, when we run the simulation for estimating gas, we consume the capsule
+  // that was meant for the actual call. So we need to push it again here. Hopefully this design
+  // will go away soon.
+  await this.getDeploymentFunctionCalls(options);
+  request.fee = await this.getFeeOptionsFromEstimatedGas(request);
+}
+```
+
+We also want to unify the mechanism for creating, simulating, proving, and sending transactions, regardless of whether they are for deploying a contract, calling a function, or deploying an account.
+
+This would clear up concerns like:
+
+> // REFACTOR: Having a `request` method with different semantics than the ones in the other
+  // derived ContractInteractions is confusing. We should unify the flow of all ContractInteractions.
+
+
+## Interface
+
+The key operations we want to perform are:
+- Simulating/Reading
+- Proving
+- Submitting
+
+These operations should be accessible from a `Wallet`, which will generally be `Schnorr` or `Signerless`. 
+
+This is analogous to `viem` having a `walletClient` and a `publicClient`.
+
+The main idea in this design is to have the BaseWallet define all the functionality to perform the key operations; the specific entrypoint in use will be able to adapt the `TxExecutionRequest`s produced to its needs.
+
+To accomplish this, we have the BaseWallet construct a `TxExecutionRequestBuilder` that can be passed to various `TxExecutionRequestAdapters`, such as one for the account entrypoint, one for the fee payment method, etc.
+
+Further, we expand `TxExecutionRequest` to allow passing things that we were previously setting directly on the PXE, so simulations can be idempotent.
+
+The interfaces are as follows:
+
+```ts
+
+// new
+export interface ArtifactAndInstance {
+  artifact: ContractArtifact;
+  instance: ContractInstanceWithAddress;
+}
+
+// modified
+export class TxExecutionRequest {
+  constructor(
+    // All these are the same:
+    public origin: AztecAddress,
+    public functionSelector: FunctionSelector,
+    public firstCallArgsHash: Fr,
+    public txContext: TxContext,
+    public argsOfCalls: PackedValues[],
+    public authWitnesses: AuthWitness[],
+    // Add:
+    /**
+     * Transient capsules needed for this execution.
+     */
+    public capsules: Fr[][],
+    /**
+     * Transient contracts needed for this execution.
+     */
+    public transientContracts: ArtifactAndInstance[],
+  ) {}
+  // ...
+}
+
+// new
+export interface InstanceDeploymentParams {
+  constructorName: string;
+  constructorArgs: any;
+  salt: Fr;
+  publicKeysHash: Fr;
```

Comment:
RE the keys instead of the hash: that is a question for @spalladino. 

RE user confusion: I'm hopeful that we can provide objects instead of arrays of arguments to cut down on that, and just have utilities that inspect the function abi to convert the objects into arrays with the proper ordering.